api/client/client.go

```diff
@@ -55,11 +58,28 @@ func (t TokenAuth) RequireTransportSecurity() bool {
 }
 
 // NewClient creates a client.
-func NewClient(ma multiaddr.Multiaddr, opts ...grpc.DialOption) (*Client, error) {
+func NewClient(ma multiaddr.Multiaddr, optsOverrides ...grpc.DialOption) (*Client, error) {
 	addr, err := util.TCPAddrFromMultiAddr(ma)
 	if err != nil {
 		return nil, err
 	}
+
+	var creds credentials.TransportCredentials
+	if strings.Contains(addr, "443") {
+		creds = credentials.NewTLS(&tls.Config{})
+	}
+
+	var opts []grpc.DialOption
+	if creds != nil {
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else {
+		opts = append(opts, grpc.WithInsecure())
+	}
+
+	auth := TokenAuth{}
+	opts = append(opts, grpc.WithPerRPCCredentials(auth))
+	opts = append(opts, optsOverrides...)
```

Comment:
Quick look at the gRPC options functions looks as though any option run later will override whatever the previous one did, so appending the overrides after the defaults here.

Side note, I love the functional options stuff, but the pattern does leave a lot of unanswered questions about how things behave in cases like providing the same option twice. Fun stuff.